### DISCUSSION TITLE
Looking for way to capture a 'window' of ADC values at high frequency

### DISCUSSION THREAD

**Discussion by veebch** [micropython/micropython#13169] [RP2040 / Pico] [answered]
I'm building a flash meter and have a nice fast light sensor (photodiode and transimpedence amp.)

I want to find a way to monitor the ADC at as close to the 500ksps as I can get. A typical flash duration is about 1/1000th sec, with the quick ones being around 1/5000th sec.

There are examples of how to do this in C, but given that my code is currently in upython,  I'd like to know if there are any options I should pursue before switching to C. A viper emitter seems to get me to 160 ksps which *might* be fast enough. 

Anyway my question is: (How) can I sample the ADC at close to 500ksps?

**Comment by 2dof** [comment]
If flash duration is 1/5000th sec i.e. pulse corresponds to 5kHz frequency, then depending on how detailed you want to sample a pulse envelope then usually 5 to 10 x 5kHz  = 25 to 50 kHz sampling should be OK, so  "A viper emitter seems to get me to 160 ksps." will be enough

But if you want "catch" typical optical pulse radiation (laser/ fast flash) is about 80 us with 4-6 us for rising slope then 160ksps will be slow for catching rising slope.

**Reply by GitHubsSilverBullet** [reply]
Easy!
Set the ADC to continuous mode and either use DMA or a few lines of Thumb Assembler to read the FIFO.
You'll get the full  500ksps, no problem at all..
If you insist on writing to flash, delegate that to the second core. Wouldn't be surprised if the PI flash won't live very long.
Instead, if you expect a slightly slower changing input, safe the difference between two samples as a signed int8.
Save that to RAM, there's quite al lot there, at least for 1/3 of a second.

**Reply by mendenm** [reply]
yes, this is a perfect use case for DMA, which will perfectly capture at the full speed of the ADC.  Since the flash is very short, you don't even need a huge buffer to capture it.  You do need some type of trigger to start the acquisition before the flash.  DMA can be set up nicely in micropython.

**Reply by veebch** [reply]
My pre-dma solution is to loop through an writing ADC values on an infinite loop, the break condition is when the light drops by 95 percent. That means that the you capture the light levels pre-and post-flash. full flash is in that window.

The resolution still looks OK, but it's only running at about 100ksps, so DMA it is!

**Reply by mendenm** [reply]
you can ping-pong or multiple buffer in DMA using the DMA chaining feature, and then do a fast  scan of the buffers in real time to see if you have a trigger, then stop the DMA

**Comment by veebch** [comment]
@GitHubsSilverBullet 

> Easy!

Hooray!

> Set the ADC to continuous mode and either use DMA or a few lines of Thumb Assembler to read the FIFO.
> You'll get the full 500ksps, no problem at all..

Fires up google

Thanks

**Reply by veebch** [reply]
@GitHubsSilverBullet 

After several hours of googleage, I am requesting a smidge more info to get me going :)

**Reply by GitHubsSilverBullet** [reply, marked as answer]
Allocate an array of sufficient size of 8bit or 16bit unsigned values. (import array)
Use register modification (datasheet) to:
- Enable ADC on required channel.
- Set max ADC speed. Divider INT/FRAC. 
- Change ADC to continuous mode.
- Enable ADC FIFO.

It now spews out a continuous stream of ADC values via the FIFO.

Write a function (viper will do) which constantly checks the FIFO count,  reads the values and puts'em into the (indexed) array. 
That's basically it.
You'll have to think about trigger conditions, either at the beginning of after  the measurement.
You'll also have to hope that a 2µs resolution is quick enough.

In case a pure digital (1/0) decision is enough, use the PIO instead. It'll provide a 8ns resolution @ standard CPU speed. The Schmitt-Trigger has a ~ 200mV hysteresis around ~ 1.65V

**Reply by veebch** [reply]
Thankyou!
